Document/0x05c-Reverse-Engineering-and-Tampering.md

```diff
@@ -444,14 +477,24 @@ BX   R2
 
 When this function returns, R0 contains a pointer to the newly constructed UTF string. This is the final return value, so R0 is left unchanged and the function returns.
 
+
+###### Ghidra
+
+On opening the library in Ghidra we can see all the function define in the **Symbol Tree** in **Functions** tree. The native library for the current application is relatively very small. There are three user defined functions - `FUN_001004d0`, `FUN_0010051c`, and `Java_sg_vantagepoint_helloworldjni_MainActivity_stringFromJNI`. The other symbols are not user defined and are generated for proper functioning of the shared library. The instructions in the function `Java_sg_vantagepoint_helloworldjni_MainActivity_stringFromJNI` are already discussed in detail in previous sections. In this section we can look into the decompilation of the function. 
+
+Inside the current function there is a call to another function, whose address is obtained by accessing an offset in JNIEnv pointer. This logic has been diagrammatically demonstrated above as well. Corresponding C code for the disassembled function is shown in the decompilation window. This decompiled C code makes it much easier to understand the function call made. Since this function is small and not complex, the decompilation output is accurate, this can change drastically when dealing with complex functions.
```

Comment:
```suggestion
Inside the current function there is a call to another function, whose address is obtained by accessing an offset in the `JNIEnv` pointer (found as `plParm1`). This logic has been diagrammatically demonstrated above as well. The corresponding C code for the disassembled function is shown in the **Decompiler** window. This decompiled C code makes it much easier to understand the function call being made. Since this function is small and extremely simple, the decompilation output is very accurate, this can change drastically when dealing with complex functions.
```

Reply:
Is the `JNIEnv` pointer explained before? If not could you make the user understand this? This is a very interesting point to highlight when reversing Android native libs. I've just searched the chapter for JNIEnv and it's mentioned but not explained. Maybe you could take the chance to explain it in the first location where it's being referred (around line 180?). That would be very helpful so that the reader can understand what's behind it and why is this offset `0x538` being added to it ;)

Reply:
Sure, will fix it.